kong/router/fields.lua

```diff
@@ -0,0 +1,317 @@
+    function(v, c, params)
+      local queries = params.queries
+      if not queries then
+        return true
+      end
+
+      for _, n in ipairs(v) do
+        local v = queries[n]
+        local f = "http.queries." .. n
+
+        -- the query parameter has only one value, like /?foo=bar
+        if type(v) == "string" then
+          local res, err = c:add_value(f, v)
+          if not res then
+            return nil, err
+          end
+
+        -- the query parameter has no value, like /?foo,
+        -- get_uri_arg will get a boolean `true`
+        -- we think it is equivalent to /?foo=
+        elseif type(v) == "boolean" then
+          local res, err = c:add_value(f, "")
+          if not res then
+            return nil, err
+          end
+
+        -- multiple values for a single query parameter, like /?foo=bar&foo=baz
+        elseif type(v) == "table" then
+          for _, v in ipairs(v) do
+            local res, err = c:add_value(f, v)
+            if not res then
+              return nil, err
+            end
+          end
+        end -- if type(v)
+      end   -- for ipairs(v)
+
+      return true
+    end,
+}
+
+
+function get_cache_key(fields, params)
+  for field, func in pairs(CACHE_KEY_FUNCS) do
+    local value = fields[field]
+
+    if value or                 -- true or table
+       field == "http.host" or  -- preserve_host
+       field == "http.path"     -- 05-proxy/02-router_spec.lua:1329
+    then
+      func(value, params, str_buf)
+    end
+  end
+
+  return str_buf:get()
+end
+
+
+else -- stream subsystem
+
+
+CACHE_KEY_FUNCS = {
+    ["net.src.ip"] =
+    function(v, params, buf)
+      buf:put(params.src_ip or ""):put("|")
+    end,
+
+    ["net.src.port"] =
+    function(v, params, buf)
+      buf:put(params.src_port or ""):put("|")
+    end,
+
+    ["net.dst.ip"] =
+    function(v, params, buf)
+      buf:put(params.dst_ip or ""):put("|")
+    end,
+
+    ["net.dst.port"] =
+    function(v, params, buf)
+      buf:put(params.dst_port or ""):put("|")
+    end,
+
+    ["tls.sni"] =
+    function(v, params, buf)
+      buf:put(params.sni or ""):put("|")
+    end,
+}
+
+
+MATCH_CTX_FUNCS = {
+    ["net.src.ip"] =
+    function(v, c, params)
+      return c:add_value("net.src.ip", params.src_ip)
+    end,
+
+    ["net.src.port"] =
+    function(v, c, params)
+      return c:add_value("net.src.port", params.src_port)
+    end,
+
+    ["net.dst.ip"] =
+    function(v, c, params)
+      return c:add_value("net.dst.ip", params.dst_ip)
+    end,
+
+    ["net.dst.port"] =
+    function(v, c, params)
+      return c:add_value("net.dst.port", params.dst_port)
+    end,
+
+    ["tls.sni"] =
+    function(v, c, params)
+      return c:add_value("tls.sni", params.sni)
+    end,
+
+    ["net.protocol"] =
+    function(v, c, params)
+      return c:add_value("net.protocol", params.scheme)
+    end,
+}
+
+
+function get_cache_key(fields, params)
+  for field, func in pairs(CACHE_KEY_FUNCS) do
```

Comment:
Why are we iterating `CACHE_KEY_FUNCS` instead of doing this?

```lua
for field, value in pairs(fields) do
  handle_func[field](str_buf, value)
end
```

If the order matters, please refer to this: https://luajit.org/faq.html#:~:text=Q%3A%20Table%20iteration%20with%20pairs()%20does%20not%20result%20in%20the%20same%20order%3F

We should not rely on the behavior of pairs.

Reply:
changed. Notice: we must use `http.host` and `http.path` to calculate cache key, or else some test cases will fail.

Reply:
discuss in https://github.com/Kong/kong/pull/12127#discussion_r1430822265